src/realm/alloc_slab.cpp

```diff
@@ -338,6 +338,10 @@ MemRef SlabAlloc::do_alloc(const size_t size)
     // Round upwards to nearest page size
     new_size = ((new_size - 1) | (page_size() - 1)) + 1;
 
+    if (ref + new_size > maximum_realm_size) {
+        throw std::runtime_error("Realm size exceeded " + realm::util::to_string(maximum_realm_size) + " bytes");
```

Comment:
One solution could be to close and compact?
If this is actually a problem it is probably due to a poorly designed app, either storing large binary data or pinning old versions.
However core should be doing more to prevent this and solving #2343 would certainly help.